types/node/globals.d.ts

```diff
@@ -53,10 +53,34 @@ declare global {
     // Same as module.exports
     var exports: any;
 
+    interface GCFunction {
+        (options: {
+            execution?: "sync";
+            flavor?: "regular" | "last-resort";
+            type?: "major-snapshot" | "major" | "minor";
+            filename?: string;
+        }): void;
+        (options: {
+            execution?: "async";
+            flavor?: "regular" | "last-resort";
+            type?: "major-snapshot" | "major" | "minor";
+            filename?: string;
+        }): Promise<void>;
+        (options?: boolean): void;
+        (
+            options?: {
+                execution?: "async" | "sync";
+                flavor?: "regular" | "last-resort";
+                type?: "major-snapshot" | "major" | "minor";
+                filename?: string;
+            } | boolean,
+        ): void | Promise<void>;
```

Comment:
```suggestion
```
Looks like an implementation signature, which isn't required here.

Reply:
I'm not sure if it's required or not. If I omit it, then this code fails to typecheck:

```tsx
function foo(async?:boolean){
  return gc({
    execution: async ? "async":"sync"
  })
}
```

[Playground Link](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgOIGEBiBXECzAD2IyA3gLABQyNyAFIQA4HEDOAXGVbT8hAB4QE2FiAD8nAESsAnnkkBubrxowANnABuhKBOSSoEAObYNUScgA++jazABaQ6x1hFylWBmMIeyQFs4ACsde1YQOEZWAAtCVyt9AODzeP9QHTdqFWQYYDUIcL8fTjsoUCMlTJoAXwBKTm1gABMKngZmIhAOLkreASERDqk4WXkWrPUtHV9DEzMLa0lbBycXDKzkT29fRJCwiOjY+YSg9JS-NPMxlRy8gqLkErKr5FrOAAUoQnPWCAAeBsaAD5nm1RKw9AAjQiEPJwEB1ZAAipVKhURpCMwoBBsMDIIwIdgYHB4USoygwXD4DrZaF0YZyBASKEwiBwmoUTKGMDYKAkfF0Dk8PrCUSc

https://github.com/microsoft/TypeScript/issues/14107

Reply:
Not sure what the usage case would be, but if you want to allow `execution` to be a union while keeping the return type dependent on its value, you'd need to use a conditional return. I guess that would look something like:

```ts
    interface GCFunction {
        <ExecutionOption extends "sync" | "async" | undefined>(options: {
            execution: ExecutionOption;
            flavor?: "regular" | "last-resort" | undefined;
            type?: "major-snapshot" | "major" | "minor" | undefined;
            filename?: string | undefined;
        }): ExecutionOption extends "async" ? Promise<void> : void;
        (options: {
            flavor?: "regular" | "last-resort" | undefined;
            type?: "major-snapshot" | "major" | "minor" | undefined;
            filename?: string | undefined;
        }): void;
        (options?: boolean): void;
    }
```

Reply:
Aye, this wouldn't allow a union type for `execution`, although I don't really see the usage case.

If you wanted this, you'd need a conditional return, something like:

```ts
    interface GCFunction {
        <ExecutionOption extends "sync" | "async" | undefined>(options: {
            execution: ExecutionOption;
            flavor?: "regular" | "last-resort" | undefined;
            type?: "major-snapshot" | "major" | "minor" | undefined;
            filename?: string | undefined;
        }): ExecutionOption extends "async" ? Promise<void> : void;
        (options: {
            flavor?: "regular" | "last-resort" | undefined;
            type?: "major-snapshot" | "major" | "minor" | undefined;
            filename?: string | undefined;
        }): void;
        (options?: boolean): void;
    }
```

Reply:
Okay. I'm happy enough with the current implementation and the conditional return type is ugly.

It seriously bugs me that `ReturnType<typeof gc>` and `Parameters<typeof gc>` don't respect the overloading. But I'm going to chalk this up to buggy behavior of TypeScript itself and flog on the upstream issues.